### PR TITLE
KAFKA-15022: introduce interface to control graph constructor

### PR DESCRIPTION
**Pull request by lihaosky**
### Description
Refactor graph construction and assignment in `RackAwareAssignor` to new interface. Will add implementation for subtopology case and unit test later. This is on top of https://github.com/apache/kafka/pull/14711

### CODE REVIEW

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/RackAwareGraphConstructor.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+    int getSinkNodeID(final List<TaskId> taskIdList, final List<UUID> clientList, final Map<Subtopology, Set<TaskId>> tasksForTopicGroup);
+
+    int getClientNodeId(final int clientIndex, final List<TaskId> taskIdList, final List<UUID> clientList, final int topicGroupIndex);
+
+    int getClientIndex(final int clientNodeId, final List<TaskId> taskIdList, final List<UUID> clientList, final int topicGroupIndex);
+
+    Graph<Integer> constructTaskGraph(final List<UUID> clientList,
+        final List<TaskId> taskIdList,
+        final SortedMap<UUID, ClientState> clientStates,
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/RackAwareGraphConstructor.java]: Thia was a `Map` in the old code -- why the change to `SortedMap` ?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/RackAwareGraphConstructor.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    int SOURCE_ID = -1;
+
+    int getSinkNodeID(final List<TaskId> taskIdList, final List<UUID> clientList, final Map<Subtopology, Set<TaskId>> tasksForTopicGroup);
+
+    int getClientNodeId(final int clientIndex, final List<TaskId> taskIdList, final List<UUID> clientList, final int topicGroupIndex);
+
+    int getClientIndex(final int clientNodeId, final List<TaskId> taskIdList, final List<UUID> clientList, final int topicGroupIndex);
+
+    Graph<Integer> constructTaskGraph(final List<UUID> clientList,
+        final List<TaskId> taskIdList,
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/RackAwareGraphConstructor.java:47]: nit: align indention (same below)

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/MinTrafficGraphConstructor.java [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+import java.util.function.BiPredicate;
+import org.apache.kafka.streams.processor.TaskId;
+import org.apache.kafka.streams.processor.internals.TopologyMetadata.Subtopology;
+import org.apache.kafka.streams.processor.internals.assignment.RackAwareTaskAssignor.GetCostFunction;
+
+public class MinTrafficGraphConstructor implements RackAwareGraphConstructor {
+
+    @Override
+    public int getSinkNodeID(final List<TaskId> taskIdList, final List<UUID> clientList,
+        final Map<Subtopology, Set<TaskId>> tasksForTopicGroup) {
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/MinTrafficGraphConstructor.java]: nit: formatting
```
// either
public int getSinkNodeID(final List<TaskId> taskIdList, final List<UUID> clientList, final Map<Subtopology, Set<TaskId>> tasksForTopicGroup) {

// or
public int getSinkNodeID(
    final List<TaskId> taskIdList,
    final List<UUID> clientList,
    final Map<Subtopology, Set<TaskId>> tasksForTopicGroup
) {
```

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/MinTrafficGraphConstructor.java [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+    public int getClientIndex(final int clientNodeId, final List<TaskId> taskIdList, final List<UUID> clientList, final int topicGroupIndex) {
+        return clientNodeId - taskIdList.size();
+    }
+
+    @Override
+    public Graph<Integer> constructTaskGraph(final List<UUID> clientList,
+        final List<TaskId> taskIdList, final SortedMap<UUID, ClientState> clientStates,
+        final Map<TaskId, UUID> taskClientMap, final Map<UUID, Integer> originalAssignedTaskNumber,
+        final BiPredicate<ClientState, TaskId> hasAssignedTask, final GetCostFunction getCostFunction, final int trafficCost,
+        final int nonOverlapCost, final boolean hasReplica, final boolean isStandby) {
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/MinTrafficGraphConstructor.java]: nit: formatting (one parameter per line -- cf comment above)

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/MinTrafficGraphConstructor.java [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+                    originalAssignedTaskNumber.merge(clientState.getKey(), 1, Integer::sum);
+                }
+            }
+        }
+
+        // Make task and client Node id in graph deterministic
+        for (int taskNodeId = 0; taskNodeId < taskIdList.size(); taskNodeId++) {
+            final TaskId taskId = taskIdList.get(taskNodeId);
+            for (int j = 0; j < clientList.size(); j++) {
+                final int clientNodeId = getClientNodeId(j, taskIdList, clientList, 0);
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/MinTrafficGraphConstructor.java]: Given that this is min-cost case, so we need to pass `clientList` or could we pass `null` (similar to calling `getSinkNodeID(...)` with `null` below) ?

Should we pass `0` or `-1` to highly "unused" better?

Might be worth to add a comment?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/MinTrafficGraphConstructor.java [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+
+            // Add edge from source to task
+            graph.addEdge(SOURCE_ID, taskNodeId, 1, 0, 1);
+        }
+
+        final int sinkId = getSinkNodeID(taskIdList, clientList, null);
+        // It's possible that some clients have 0 task assign. These clients will have 0 tasks assigned
+        // even though it may have higher traffic cost. This is to maintain the original assigned task count
+        for (int i = 0; i < clientList.size(); i++) {
+            final int clientNodeId = getClientNodeId(i, taskIdList, clientList, 0);
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/MinTrafficGraphConstructor.java]: as above?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/MinTrafficGraphConstructor.java [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+    }
+
+    @Override
+    public boolean assignTaskFromMinCostFlow(final Graph<Integer> graph,
+        final List<UUID> clientList, final List<TaskId> taskIdList,
+        final Map<UUID, ClientState> clientStates,
+        final Map<UUID, Integer> originalAssignedTaskNumber, final Map<TaskId, UUID> taskClientMap,
+        final BiConsumer<ClientState, TaskId> assignTask,
+        final BiConsumer<ClientState, TaskId> unAssignTask,
+        final BiPredicate<ClientState, TaskId> hasAssignedTask) {
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/MinTrafficGraphConstructor.java]: nit: fix indention (cf other comments).

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/MinTrafficGraphConstructor.java [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+
+        int tasksAssigned = 0;
+        boolean taskMoved = false;
+        for (int taskNodeId = 0; taskNodeId < taskIdList.size(); taskNodeId++) {
+            final TaskId taskId = taskIdList.get(taskNodeId);
+            final Map<Integer, Graph<Integer>.Edge> edges = graph.edges(taskNodeId);
+            for (final Graph<Integer>.Edge edge : edges.values()) {
+                if (edge.flow > 0) {
+                    tasksAssigned++;
+                    final int clientIndex = getClientIndex(edge.destination, taskIdList, clientList, 0);
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/MinTrafficGraphConstructor.java]: as above?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/RackAwareTaskAssignor.java [hunk 9]
```diff
@@ -363,13 +361,14 @@ public long optimizeActiveTasks(final SortedSet<TaskId> activeTasks,
         final List<TaskId> taskIdList = new ArrayList<>(activeTasks);
         final Map<TaskId, UUID> taskClientMap = new HashMap<>();
         final Map<UUID, Integer> originalAssignedTaskNumber = new HashMap<>();
-        final Graph<Integer> graph = constructTaskGraph(clientList, taskIdList,
-            clientStates, taskClientMap, originalAssignedTaskNumber, ClientState::hasActiveTask, trafficCost, nonOverlapCost, false, false);
+        final RackAwareGraphConstructor graphConstructor = RackAwareGraphConstructorFactory.create(assignmentConfigs, tasksForTopicGroup);
+        final Graph<Integer> graph = graphConstructor.constructTaskGraph(clientList, taskIdList,
+            clientStates, taskClientMap, originalAssignedTaskNumber, ClientState::hasActiveTask, this::getCost, trafficCost, nonOverlapCost, false, false);
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/RackAwareTaskAssignor.java]: nit: formatting (pass one parameter per line):
```
final Graph<Integer> graph = graphConstructor.constructTaskGraph(
    clientList,
    taskIdList,
    ...,
    false,
    false
);
```

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/RackAwareTaskAssignor.java [hunk 10]
```diff
@@ -327,8 +325,8 @@ private long tasksCost(final SortedSet<TaskId> tasks,
         }
         final List<UUID> clientList = new ArrayList<>(clientStates.keySet());
         final List<TaskId> taskIdList = new ArrayList<>(tasks);
-        final Graph<Integer> graph = constructTaskGraph(clientList, taskIdList,
-            clientStates, new HashMap<>(), new HashMap<>(), hasAssignedTask, trafficCost, nonOverlapCost, hasReplica, isStandby);
+        final Graph<Integer> graph = RackAwareGraphConstructorFactory.create(assignmentConfigs, tasksForTopicGroup).constructTaskGraph(clientList, taskIdList,
+            clientStates, new HashMap<>(), new HashMap<>(), hasAssignedTask, this::getCost, trafficCost, nonOverlapCost, hasReplica, isStandby);
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/RackAwareTaskAssignor.java]: nit: formatting (pass one parameter per line):
```
final Graph<Integer> graph = RackAwareGraphConstructorFactory
    .create(assignmentConfigs, tasksForTopicGroup)
    .constructTaskGraph(
        clientList,
        taskIdList,
        ...,
        hasReplica,
        isStandby
    );
````

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/RackAwareTaskAssignor.java [hunk 11]
```diff
@@ -61,19 +59,30 @@ boolean canMove(final ClientState source,
                         final Map<UUID, ClientState> clientStateMap);
     }
 
+    @FunctionalInterface
+    public interface GetCostFunction {
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/RackAwareTaskAssignor.java]: `GetCostFunction` -> `CostFunction`

(This interface does not _return_ a cost function, but _is_ a cost function, right?)

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/BalanceSubtopologyGraphConstructor.java [hunk 12]
```diff
@@ -0,0 +1,186 @@
...
+
+    private final Map<Subtopology, Set<TaskId>> tasksForTopicGroup;
+
+    public BalanceSubtopologyGraphConstructor(final Map<Subtopology, Set<TaskId>> tasksForTopicGroup) {
+        this.tasksForTopicGroup = tasksForTopicGroup;
+    }
+
+    @Override
+    public int getSinkNodeID(final List<TaskId> taskIdList, final List<UUID> clientList,
+        final Map<Subtopology, Set<TaskId>> tasksForTopicGroup) {
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/BalanceSubtopologyGraphConstructor.java]: formatting (cf other comments)

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/BalanceSubtopologyGraphConstructor.java [hunk 13]
```diff
@@ -0,0 +1,186 @@
...
+    private static int getSecondStageClientNodeId(final List<TaskId> taskIdList, final List<UUID> clientList, final Map<Subtopology, Set<TaskId>> tasksForTopicGroup, final int clientIndex) {
+        return taskIdList.size() + clientList.size() * tasksForTopicGroup.size() + clientIndex;
+    }
+
+    @Override
+    public Graph<Integer> constructTaskGraph(final List<UUID> clientList,
+        final List<TaskId> taskIdList, final SortedMap<UUID, ClientState> clientStates,
+        final Map<TaskId, UUID> taskClientMap, final Map<UUID, Integer> originalAssignedTaskNumber,
+        final BiPredicate<ClientState, TaskId> hasAssignedTask, final GetCostFunction getCostFunction, final int trafficCost,
+        final int nonOverlapCost, final boolean hasReplica, final boolean isStandby) {
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/BalanceSubtopologyGraphConstructor.java]: nit: formatting

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/BalanceSubtopologyGraphConstructor.java [hunk 14]
```diff
@@ -0,0 +1,186 @@
...
+        for (final Entry<Subtopology, Set<TaskId>> kv : sortedTasksForTopicGroup.entrySet()) {
+            final SortedSet<TaskId> taskIds = new TreeSet<>(kv.getValue());
+            for (int clientIndex = 0; clientIndex < clientList.size(); clientIndex++) {
+                final UUID processId = clientList.get(clientIndex);
+                final int clientNodeId = getClientNodeId(clientIndex, taskIdList, clientList, topicGroupIndex);
+                int startingTaskNodeId = taskNodeId;
+                for (final TaskId taskId : taskIds) {
+                    final int flow = hasAssignedTask.test(clientStates.get(processId), taskId) ? 1 : 0;
+                    graph.addEdge(startingTaskNodeId, clientNodeId, 1, getCostFunction.getCost(taskId, processId, false, trafficCost, nonOverlapCost, isStandby), flow);
+                    graph.addEdge(SOURCE_ID, startingTaskNodeId, 1, 0, 0);
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/BalanceSubtopologyGraphConstructor.java]: It seems we add this edge multiple times?

Would it be better to flip the nesting of `for (final TaskId taskId : taskIds)` and `for (int clientIndex = 0; clientIndex < clientList.size(); clientIndex++) {` to allow us to assign this edge only once?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/BalanceSubtopologyGraphConstructor.java [hunk 15]
```diff
@@ -0,0 +1,186 @@
...
+                    startingTaskNodeId++;
+                }
+
+                final int secondStageClientNodeId = getSecondStageClientNodeId(taskIdList, clientList, tasksForTopicGroup, clientIndex);
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/BalanceSubtopologyGraphConstructor.java]: Seems that if we flip the nesting, we move the issue to here? -- Mabye you try to do to many things at one in the same loops? Would it be better to break it a little bit better, and loop over all sub-topology two times (instead of ones?)

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/BalanceSubtopologyGraphConstructor.java [hunk 16]
```diff
@@ -0,0 +1,186 @@
...
+    }
+
+    @Override
+    public boolean assignTaskFromMinCostFlow(final Graph<Integer> graph,
+        final List<UUID> clientList, final List<TaskId> taskIdList,
+        final Map<UUID, ClientState> clientStates,
+        final Map<UUID, Integer> originalAssignedTaskNumber, final Map<TaskId, UUID> taskClientMap,
+        final BiConsumer<ClientState, TaskId> assignTask,
+        final BiConsumer<ClientState, TaskId> unAssignTask,
+        final BiPredicate<ClientState, TaskId> hasAssignedTask) {
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/BalanceSubtopologyGraphConstructor.java]: nit: formatting

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/BalanceSubtopologyGraphConstructor.java [hunk 17]
```diff
@@ -0,0 +1,186 @@
...
+
+        int taskNodeId = 0;
+        int topicGroupIndex = 0;
+        int tasksAssigned = 0;
+        boolean taskMoved = false;
+        for (final Entry<Subtopology, Set<TaskId>> kv : sortedTasksForTopicGroup.entrySet()) {
+            final SortedSet<TaskId> taskIds = new TreeSet<>(kv.getValue());
+            for (final TaskId taskId : taskIds) {
+                final Map<Integer, Graph<Integer>.Edge> edges = graph.edges(taskNodeId);
+                for (final Graph<Integer>.Edge edge : edges.values()) {
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/BalanceSubtopologyGraphConstructor.java]: Seems this `for` loop is the same as for min-cost? Should we extract the code into a helper method?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/BalanceSubtopologyGraphConstructor.java [hunk 18]
```diff
@@ -0,0 +1,186 @@
...
+                        assignTask.accept(clientStates.get(processId), taskId);
+                        taskMoved = true;
+                    }
+                }
+                taskNodeId++;
+            }
+            topicGroupIndex++;
+        }
+
+        // Validate task assigned
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/BalanceSubtopologyGraphConstructor.java]: Seems this validation code (starting here to the end of the method) is the same as for min-traffic? Should we extract a helper method and share code?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/Graph.java [hunk 19]
```diff
@@ -220,6 +222,78 @@ public void solveMinCostFlow() {
...
+                final V parent = parents.get(node);
+                adjList.get(parent).get(node).residualFlow -= flow;
+                adjList.get(node).get(parent).residualFlow += flow;
+            }
+
+            parents = new HashMap<>();
+        }
+    }
+
+    private boolean bfs(final V source, final V target, final Map<V, V> parents) {
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/Graph.java]: `bfs` -> avoid abbreviations

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/BalanceSubtopologyGraphConstructor.java (hunk 15 above)
Comment (lihaosky) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/BalanceSubtopologyGraphConstructor.java]: Right, I initially put task in outer loop and found this issue. Will make it more readable.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/assignment/AssignmentTestUtils.java [hunk 20]
```diff
@@ -676,18 +676,28 @@ static SortedMap<TaskId, Set<TopicPartition>> getTaskTopicPartitionMap(final int
...
-    static Map<String, Object> configProps(final boolean enableRackAwareAssignor, final int replicaNum) {
+    static Map<String, Object> configProps(final String rackAwareConfig, final int replicaNum) {
         final Map<String, Object> configurationMap = new HashMap<>();
         configurationMap.put(StreamsConfig.APPLICATION_ID_CONFIG, APPLICATION_ID);
         configurationMap.put(StreamsConfig.BOOTSTRAP_SERVERS_CONFIG, USER_END_POINT);
         configurationMap.put(StreamsConfig.NUM_STANDBY_REPLICAS_CONFIG, replicaNum);
-        if (enableRackAwareAssignor) {
-            configurationMap.put(StreamsConfig.RACK_AWARE_ASSIGNMENT_STRATEGY_CONFIG, StreamsConfig.RACK_AWARE_ASSIGNMENT_STRATEGY_MIN_TRAFFIC);
-        }
+        // configurationMap.put(StreamsConfig.RACK_AWARE_ASSIGNMENT_STRATEGY_CONFIG, StreamsConfig.RACK_AWARE_ASSIGNMENT_STRATEGY_MIN_TRAFFIC);
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/assignment/AssignmentTestUtils.java]: Needs some cleanup

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/assignment/GraphTest.java [hunk 21]
```diff
@@ -394,6 +394,106 @@ public void testDeterministic() {
         }
     }
 
+    @Test
+    public void testMaxFlowOnlySourceAndSink() {
+        final Graph<Integer> graph1 = new Graph<>();
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/assignment/GraphTest.java:399]: Nit: Why `graph1` but not `graph`? (similar below)

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/assignment/RackAwareGraphConstructorTest.java [hunk 22]
```diff
@@ -0,0 +1,291 @@
...
+        while (totalAssigned < taskIdList.size()) {
+            for (final ClientState clientState : clientStateMap.values()) {
+                if (AssignmentTestUtils.getRandom().nextInt(3) == 0) {
+                    clientState.assignActive(taskIdList.get(totalAssigned));
+                    totalAssigned++;
+                    if (totalAssigned >= taskIdList.size()) {
+                        break;
+                    }
+                } else {
+                    System.out.println();
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/assignment/RackAwareGraphConstructorTest.java]: need cleanup?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/BalanceSubtopologyGraphConstructor.java [hunk 23]
```diff
@@ -88,55 +93,20 @@ public Graph<Integer> constructTaskGraph(
             }
         }
 
-        // TODO: validate tasks in tasksForTopicGroup and taskIdList
-        final SortedMap<Subtopology, Set<TaskId>> sortedTasksForTopicGroup = new TreeMap<>(tasksForTopicGroup);
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/BalanceSubtopologyGraphConstructor.java]: Was this code only moved, or altered?

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/assignment/RackAwareGraphConstructorTest.java [hunk 24]
```diff
@@ -97,6 +98,17 @@ private int getCost(final TaskId taskId, final UUID processId, final boolean inC
         return 1;
     }
 
+    @Test
+    public void testSubtopicShouldContainAllTasks() {
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/assignment/RackAwareGraphConstructorTest.java]: `Subtopology` instead of `Subtopic` ?

**Review by mjsax — COMMENTED** [review]
Few nits / questions.

Also build failed:
```
Task :streams:compileJava

/home/jenkins/jenkins-agent/workspace/Kafka_kafka-pr_PR-14714/streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/BalanceSubtopologyGraphConstructor.java:226: warning: [lossy-conversions] implicit cast from long to int in compound assignment is possibly lossy

            taskNodeId += taskIds.stream().filter(taskIdSet::contains).count();
```

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/assignment/GraphTest.java (hunk 21 above)
Comment (lihaosky) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/assignment/GraphTest.java:399]: `graph` was used in setup

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/BalanceSubtopologyGraphConstructor.java (hunk 23 above)
Comment (lihaosky) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/BalanceSubtopologyGraphConstructor.java]: It's put in `validateTasks`

**Review by mjsax — APPROVED** [review]
LGTM.

Build failed (reason unclear). Did retrigger Jenkins. Can merge after we got a good build.

**Comment by mjsax** [comment]
Jenkins issues:
- build 7: `JDK 11 and Scala 2.13` failed
- build 6: `JDK 8 and Scala 2.12` failed

Test failures seems to be unrelated to this PR. Merging.
